src/signature/commitment_signature.rs

```diff
@@ -0,0 +1,292 @@
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+//! Digital Signature module
+//! This module defines generic traits for handling the digital signature operations, agnostic
+//! of the underlying elliptic curve implementation
+
+use crate::{
+    commitment::{HomomorphicCommitment, HomomorphicCommitmentFactory},
+    keys::{PublicKey, SecretKey},
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    cmp::Ordering,
+    hash::{Hash, Hasher},
+    marker::PhantomData,
+    ops::{Add, Mul},
+};
+use tari_utilities::{hex::Hex, ByteArray};
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error, PartialEq, Eq, Deserialize, Serialize)]
+pub enum CommitmentSignatureError {
+    #[error("An invalid challenge was provided")]
+    InvalidChallenge,
+}
+
+#[allow(non_snake_case)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CommitmentSignature<P, K, C> {
+    public_commitment_nonce: HomomorphicCommitment<P>,
+    signature_u: K,
+    signature_v: K,
+    _commitment_factory: PhantomData<C>,
+}
+
+impl<P, K, C> CommitmentSignature<P, K, C>
+where
+    P: PublicKey<K = K>,
+    K: SecretKey,
+    C: HomomorphicCommitmentFactory<P = P> + Default,
+{
+    pub fn new(public_commitment_nonce: HomomorphicCommitment<P>, signature_u: K, signature_v: K) -> Self {
+        CommitmentSignature {
+            public_commitment_nonce,
+            signature_u,
+            signature_v,
+            _commitment_factory: PhantomData,
+        }
+    }
+
+    /// This is the left-hand side of the signature verification equation
+    pub fn calc_signature_verifier(&self) -> HomomorphicCommitment<P> {
+        // v*H + u*G = Commitment
+        let factory = C::default();
+        factory.commit(&self.signature_u, &self.signature_v)
+    }
+
+    /// Sign the provided challenge with the value commitment's value and blinding factor. The two nonces should be
+    /// completely random and never reused - that responsibility lies with the calling function.
+    ///   C = a*H + x*G          ... (Pedersen commitment to the value 'a' using blinding factor 'x')
+    ///   R = k_2*H + k_1*G      ... (a public (Pedersen) commitment nonce created with the two random nonces)
+    ///   u = k_1 + e.x          ... (the first publicly known private key of the signature signing with 'x')
+    ///   v = k_2 + e.a          ... (the second publicly known private key of the signature signing with 'a')
+    ///   signature = (R, u, v)  ... (the final signature tuple)
+    pub fn sign(
+        secret_a: K,
+        secret_x: K,
+        nonce_a: K,
+        nonce_x: K,
+        challenge: &[u8],
+    ) -> Result<Self, CommitmentSignatureError>
+    where
+        K: Add<Output = K> + Mul<P, Output = P> + Mul<Output = K>,
+    {
+        let e = match K::from_bytes(challenge) {
+            Ok(e) => e,
+            Err(_) => return Err(CommitmentSignatureError::InvalidChallenge),
+        };
+        let ea = e.clone() * secret_a;
```

Comment:
```suggestion
        let ea = &e* &secret_a;
```
will work I think, saving a clone.